pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,510 @@
+		if err != nil {
+			klog.Errorf("error processing PVs for infrastructure update: %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+// fetchAndPushPvsToQueue retrieves all PVs, filters the updatable PVs and pushes the PVs to queue worker for tags update.
+func (c *EBSVolumeTagsController) fetchAndPushPvsToQueue(infra *configv1.Infrastructure) error {
+	pvs, err := c.listPersistentVolumes()
+	if err != nil {
+		return fmt.Errorf("error fetching PVs: %v", err)
+	}
+	// Compute the hash for the new set of tags
+	newTagsHash := computeTagsHash(infra.Status.PlatformStatus.AWS.ResourceTags)
+	pvsToBeUpdated := c.filterUpdatableVolumes(pvs, newTagsHash)
+
+	// If there are no volumes to update, return early
+	if len(pvsToBeUpdated) == 0 {
+		klog.Infof("No volume tags to update as hashes are unchanged")
+		return nil
+	}
+
+	// Process the volumes in batches
+	for i := 0; i < len(pvsToBeUpdated); i += batchSize {
+		end := i + batchSize
+		if end > len(pvsToBeUpdated) {
+			end = len(pvsToBeUpdated)
+		}
+		batch := pvsToBeUpdated[i:end]
+		c.addBatchVolumesToQueueWorker(batch)
+	}
+	return nil
+}
+
+// updateEBSTags updates the tags of an AWS EBS volume with rate limiting
+func (c *EBSVolumeTagsController) updateEBSTags(ec2Client *ec2.EC2, resourceTags []configv1.AWSResourceTag,
+	pvs ...*v1.PersistentVolume) error {
+	// Prepare tags
+	tags := newAndUpdatedTags(resourceTags)
+	// Create or update the tags
+	_, err := ec2Client.CreateTags(&ec2.CreateTagsInput{
+		Resources: pvsToResourceIDs(pvs),
+		Tags:      tags,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// listPersistentVolumes lists the volume
+func (c *EBSVolumeTagsController) listPersistentVolumes() ([]*v1.PersistentVolume, error) {
+	pvList, err := c.commonClient.KubeInformers.InformersFor("").Core().V1().PersistentVolumes().Lister().List(labels.Everything())
+	if err != nil {
+		klog.Errorf("error listing volumes objects: %v", err)
+		return nil, err
+	}
+	return pvList, nil
+}
+
+// updateVolume updates the persistent volume.
+func (c *EBSVolumeTagsController) updateVolume(ctx context.Context, pv *v1.PersistentVolume) error {
+	_, err := c.commonClient.KubeClient.CoreV1().PersistentVolumes().Update(ctx, pv, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Errorf("error updating volume object %s: %v", pv.Name, err)
+		return err
+	}
+	return nil
+}
+
+// handleBatchTagUpdateFailure emits the event for tags update failure and requeue the pvNames individually.
+func (c *EBSVolumeTagsController) handleBatchTagUpdateFailure(pvList []*v1.PersistentVolume, updateErr error) {
+	// log the PVs name and add them back to queue individually
+	for _, pv := range pvList {
+		klog.Errorf("error updating volume %v tags: %v", pv.Name, updateErr)
+		c.queue.AddRateLimited(&pvUpdateQueueItem{
+			updateType: updateTypeIndividual,
+			pvNames:    convertPVsListToStringArray(pv),
+		})
+	}
+	var pvNames []string
+	for _, pv := range pvList {
+		pvNames = append(pvNames, pv.Name)
+	}
```

Comment:
Can you collapse these two loops into a single loop, and build `pvNames` at the same time as requeuing the updates?